### ISSUE TITLE
[feature] Additional arguments for hooks that deal with packages and recipes

### ISSUE THREAD

**Issue by stefansli** [conan-io/conan#12258]
I'm implementing a conan hook that deals with handling debug symbols for conan packages. For the implementation I'm often forced to dig deep into conan internals to get the information I need. The hook does some bookkeeping so it can upload debug symbols to a symbol server during the ´pre_upload_package´ hook. To do this properly I need access to the recipe revision and the package revision.
From what I can see, the package revision is missing entirely from all hooks while the recipe revision is mostly present as ´kwarg´.

Assuming revisions are supported, I'd like to have access to the package revision in the pre_upload_package and post_download_package hooks.

My current idea is to use the reference to get the metadata using the `ClientCache` helper and extract the package revision from there.

Speaking of the metadata. Can I use the metadata.json to store custom data?

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by memsharded** [comment]
In Conan 2.0 the hooks receive ``conanfile`` which can be used as ``conanfile.ref`` and ``conanfile.pref`` to get the full references, with revisions. But only the hooks that correspond to recipe methods like ``pre_build()`` and ``post_build()`` are maintained as hooks.

If we want to provide some extensions points in 2.0 to allow this, we would need to understand better what is the upload flow, what needs to be uploaded, when, in what format... Because if uploaded to a different place, the idea is that they will not be part of the package and only in the symbol server?

**Comment by stefansli** [comment]
Thanks for the response.

So the current implementation does something like this:
* in `post_package` use `dumpbin` from visual studio to locate PDBs for every `*.dll` and store the filepath for later. To be "correct" the PDBs should be associated with a specific `ref` and (ideally) `pref`.
* In `pre_upload_package` all PDBs associated with the package are copied to the symbolserver associated with the remote (configured through a conan config setting)

This approach works so far but is prone to errors since I have to carry information from the `post_package` to the `pre_upload_package` hook. This information could get out of sync for example when I download a new recipe revision.
I can circumvent this by doing everything in the `pre_upload_package` hook, but that requires reading the PDB path directly from the DLL with pure Python (since dependening on pypa packages is not ideal for hooks).
If it was possible to attach additional metadata to packages / recipes, the information could not get out of sync that easily. But the proper implementation is to do everything in `pre_upload_package` so custom metadata is not a priority.

Our approach however has a problem when using a `scratch` remote to build multiple packages on different hosts in parallel. Because I don't want to upload debug symbols to a symbol server when uploading the packages to scratch. I only want to "deploy" the debug symbols when I move the package from `scratch` to `prod` after tests have been completet.

To support this, I have to cache the PDBs somewhere (e.g. Artifactory). For this caching I need `ref` and `pref` to store the PDBs in a unique fashion. The plan is to use the `post_download_package` hook to detect when a package is downloaded from the scrach remote. If that's the case, create a marker somewhere that PDBs can be found at a specific URL. During `pre_upload_package` I check for the marker and if it exists, download PDBs from the Cache and upload them to the symbol server associated with the remote we are uploading the package to.
This "marker" could very well be a custom attribute to the package metadata kept by conan.

Sorry, this sounds very complicated but it's the best we came up with so far. Any advice is very welcome, especially regarding Conan 2.0 compatibility since we don't want to develop something that won't work in the future.

This is an overview of the current and future implementation of the hooks:
* `post_package`: Gather PDBs, store list somewhere
* `pre_upload_package`: Get PDB list (either from `post_package` or `post_download_package`), upload to symbolserver / Artifactory cache
* `pre_build`: clean PDB list for current package_id
* `post_export`: clean PDB list for current ref
* `post_download_recipe`: clean PDB list for ref

**Comment by memsharded** [comment]
Thanks very much for the detailed explanation of the use case, this is really helpful to understand it.

There is a user story that is being raised more often lately and it is the story of metadata. Lets say that the Conan packages had a way to abstract the creation and storage of metadata with the packages, in a way that they don't become integral part of the package (do not affect the package_id), and they are not retrieved by default (to avoid slow installation). Is this something that would make sense from the Conan side, and then, for the final promotion from Conan to the symbol server, that is some automation that would happen exclusively at the server side? Like "take these files from this ``pref`` metadata folder "pdbs" and copy them to this other place", implemented like a script in your CI against server APIs?

**Comment by stefansli** [comment]
If there was a way to attach custom metadata and files to a pref, this would help a lot. If it was possible to let conan handle the upload this would be even better ;)
Then all I would need to do on my side would be to attach the metadata with a hook and write a script to "promote" the PDBs from conan metadata to the symbol server.

> Is this something that would make sense from the Conan side, and then, for the final promotion from Conan to the symbol server, that is some automation that would happen exclusively at the server side?

I'm not sure if I understand correctly, but the automation would either be implemented as a hook or as a script that is run by CI or something like that. AFAIK, running the script on the server (which would be Artifactory) wouldn't be possible.
The script would ideally work with server APIs or Conan Client APIs. If there is a server API, this would be obviously preferred.
Since Artifactory cannot act as a symbol server for native code, the PDBs have to be downloaded to a CI server and then uploaded to a symbol server from there.

**Comment by memsharded** [comment]
> I'm not sure if I understand correctly, but the automation would either be implemented as a hook or as a script that is run by CI or something like that. 

Yes, it is a script or something that runs directly in the CI, not really a Conan command because it is "take this file from Artifactory and put it in a symbol server" which is an unrelated operation to Conan really. Yes, using specific server APIs, file-based most likely, I think you are right, it might be necessary to download and then upload to the symbol server, but still will be more structured than trying to automate via Conan hooks, IMO

**Comment by stefansli** [comment]
Sounds like a good idea. How do we move forward?

It looks like I have to rely on my own storage of metadata for the moment. And it seems to be safe to rely on `ClientCache.package_layout(ref)` to get `pref` and other information as input for my hook.

The metadata for packages even contain the remote the package was downloaded from, so maybe I don't need to store any information on my own.
